Support optional x-api-key header in EDS API requests

cwolfebsco opened the pull request:
### About this pull request

* adds api_key as a parameter to EDS.ini
* adds $apiKey to requests as x-api-key header if api_key is set in EDS.ini
* adds $apiKeyGuest to requests as x-api-key header, if api_key_guest is set in EDS.ini AND user is in guest mode

### Reason for Draft Pull Request

* start a conversation on how to best implement this

### About EDS API Keys

EDS API supports additional (currently undocumented) request headers for API keys. 

API keys are used to monitor access to EDS API, isolating traffic by customer and use-case.  This isolation allows EBSCO to manage performance for specific keys (use cases).  Should a request be throttled, the gateway will respond with an HTTP 429.

Existing implementations may start using API keys at any time.  EBSCO will require API keys for all customers at one point, but deadlines have not been established. Customers may have more than one API key for separate client applications or use-cases accessing EDS API. 

Note: API keys do not replace EDS API’s authentication or session tokens.

To use an API key, add a header named x-api-key to every request sent to EDS API.  Set the API key as the header’s value.

•	x-api-key: The_API_Key

cwolfebsco committed:
fix: DetailedErrorDescription not always present

With API Error 154 the DetailedErrorDescription is not always present, so setting to a hyphen when not there.

cwolfebsco committed:
add api_key to EDS.ini & Backend.php

adding a variable for EDS.ini and reading it as a variable for Backend.php

cwolfebsco committed:
Testing Debug

Testing Debug - to be removed prior to go live

cwolfebsco committed:
passing apiKey through to relevant functions

cwolfebsco committed:
removing a var_dump

cwolfebsco committed:
passing apiKey

cwolfebsco committed:
fixing missing doc parameters & whitespaces

cwolfebsco committed:
apiKey is optional

apiKey is optional so set default value
fix documentation apiKey is a string not a array

cwolfebsco committed:
fixing: too many parameters in call

cwolfebsco committed:
fix: single vs double-quotes

cwolfebsco committed:
remove a debug

this debug was helpful during initial development, but I don't think it will be that helpful in day to day operations

cwolfebsco committed:
Merge remote-tracking branch 'upstream/dev' into EDS-API-Token-support

cwolfebsco committed:
Merge remote-tracking branch 'upstream/dev' into EDS-API-Token-support

demiankatz commented on module/VuFindSearch/src/VuFindSearch/Backend/EDS/Backend.php:
```diff
@@ -202,7 +210,7 @@ public function search(
     ) {
         // process EDS API communication tokens.
-        $authenticationToken = $this->getAuthenticationToken();
+        $authenticationToken = $this->getAuthenticationToken(false);
```
Since false is the default value for the first parameter, you don't need to specify it here (unless you feel it makes things more clear).

demiankatz reviewed (changes requested):
Thanks for sharing this, @cwolfebsco!

See below for one comment on a recurring change that may not be necessary.

Beyond that, my biggest question is: did you consider just passing the API key to the constructor of the connector, instead of storing it in the backend and passing it through all the method calls? Since it is a stable value that should not change, it seems like you could greatly simplify the code by changing to that approach... but maybe there's a consideration I am overlooking that prevents that possibility.

And to be a little more explicit about my proposed approach;

The EDS connector is created here: https://github.com/vufind-org/vufind/blob/dev/module/VuFind/src/VuFind/Search/Factory/EdsBackendFactory.php#L153-L172 -- I think you could simply add the API key to the options array constructed here, and then you could process it in the constructor of \VuFindSearch\Backend\EDS\Base and use it seamlessly in all the appropriate places within that class without having to pass it through so many interfaces.

What do you think?

cwolfebsco committed:
rolling back customizations

cwolfebsco committed:
support api_key_guest

Idea is that a VuFind client can be issued two API keys. One more restrictive for Guest users and one more permissive for authenticated users

cwolfebsco committed:
add: api_key, api_key_guest and is_guest

cwolfebsco committed:
correct variable names

fixed variable names

cwolfebsco committed:
read variables from base & add to call

also add headers to "call"

cwolfebsco committed:
add: debug headers

this is helpful for checking that the various calls make the right settings

cwolfebsco committed:
fix whitespaces

cwolfebsco committed:
fix line length

cwolfebsco committed:
prefer single quotes over double

cwolfebsco committed:
trailing comma

cwolfebsco commented:
Thanks to your kind and most valuable feedback, I have been able to significantly reduce complexity. As a result I was able to add an optional use case that would see us use a different API Key for guest users versus authenticated users.

API Keys are intended, amongst other things, to help EBSCO manage API traffic. With many clients finding bot traffic on the rise, this would allow us to provision a site with two API Keys for the same EDS Profile, ensuring that authenticated users get the highest level of service, even when a guest user might see reduced service (i.e. due to increased, not yet mitigated bot traffic).

demiankatz commented on module/VuFindSearch/src/VuFindSearch/Backend/EDS/Connector.php:
```diff
@@ -101,7 +101,7 @@ protected function httpRequest(
     ) {
         $this->debug("{$method}: {$baseUrl}?{$queryString}");
-
+        $this->debug($this->varDump($headers));
```
Do we really want to debug the full header information? Probably not a huge risk of harm here, but there's at least a chance of a misconfiguration causing a credential leak if we do.

cwolfebsco replied:
You are right, this was useful during my testing, but likely won't have real benefits in production. I'll remove it for sure!

demiankatz commented on module/VuFindSearch/src/VuFindSearch/Backend/EDS/Base.php:
```diff
@@ -96,6 +96,27 @@ abstract class Base implements LoggerAwareInterface
+     *
+     * @var ?string
+     */
+    protected $apiKey;
```
Should probably explicitly default this to null since there's a chance of it being unset.

```suggestion
    protected $apiKey = null;
```

demiankatz commented on module/VuFindSearch/src/VuFindSearch/Backend/EDS/Base.php:
```diff
@@ -96,6 +96,27 @@ abstract class Base implements LoggerAwareInterface
+     *
+     * @var ?string
+     */
+    protected $apiKeyGuest;
```
Same here.

```suggestion
    protected $apiKeyGuest = null;
```

demiankatz commented on module/VuFindSearch/src/VuFindSearch/Backend/EDS/Base.php:
```diff
@@ -96,6 +96,27 @@ abstract class Base implements LoggerAwareInterface
+    /**
+     * Indicator if user "isGuest"
+     *
+     * @var ?string
+     */
+    protected $isGuest;
```
This has the wrong type and should also be initialized; I'd suggest defaulting to true on the principle of least privilege:

```suggestion
     * @var bool
     */
    protected $isGuest = true;
```

demiankatz commented on module/VuFindSearch/src/VuFindSearch/Backend/EDS/Base.php:
```diff
@@ -447,11 +478,22 @@ protected function call(
             'Accept-Encoding' => 'gzip,deflate',
         ];
+        $this->debug(
+            'isguest: ' . $this->isGuest
+            . ' | APIKey: ' . $this->apiKey
+            . ' | APIKey Guest: ' . $this->apiKeyGuest
```
Since some of these values might be non-strings, I'd suggest:

```suggestion
            'isguest: ' . ($this->isGuest ? 'true' : 'false')
            . ' | APIKey: ' . ($this->apiKey ?? '-')
            . ' | APIKey Guest: ' . ($this->apiKeyGuest ?? '-')
```

demiankatz replied:
But it might also be worth using `substr` to truncate the keys when set if you don't want to leak out the full things (though without knowing what a real one looks like, I'm not sure exactly what to suggest).

demiankatz commented on module/VuFindSearch/src/VuFindSearch/Backend/EDS/Base.php:
```diff
@@ -447,11 +478,22 @@ protected function call(
         }
+        if (!empty($this->apiKey)) {
+            $headers['x-api-key'] = $this->apiKey;
+            if ($this->isGuest && !empty($this->apiKeyGuest)) {
```
Does this really need to be nested inside the check for the regular API key? Is it possible there could be a scenario where an API key is ONLY used for guests? In that situation, this wouldn't work because the unset non-guest key would prevent the guest key from being examined at all.

maccabeelevine replied:
I think that scenario is actually likely during the initial period where the api key is optional.  Institutions may want to use the guest key for non-authenticated users in order to get that rate limiting behavior (in case of bots) while not affecting authenticated users.  I'm not saying that would work the way the customer expects -- the API may throttle everything anyway -- but I can imagine someone trying it.

cwolfebsco replied:
I'll need to ask Matt :) There are two types of key "legacy" and "new". If a _new_ key has been vended you cannot use EDS API without providing a key. I need to very my assumption that only new keys would be vended going forward and in this case you would need to have an API key and Guest API key would be optional. I'll ask Matt and adjust accordingly

demiankatz replied:
If we do end up in the scenario where guest API only works if regular API key is also set, we should add a comment to the .ini file pointing this out to prevent possible confusion in case of misconfiguration.

demiankatz reviewed (changes requested):
Thanks, @cwolfebsco, this is a big step forward... but I have a few new nitpicks for your consideration. :-)

cwolfebsco committed:
fix: remove debug of headers

this was useful during testing, but is likely not useful during production debug

cwolfebsco committed:
Update module/VuFindSearch/src/VuFindSearch/Backend/EDS/Base.php

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

cwolfebsco committed:
Update module/VuFindSearch/src/VuFindSearch/Backend/EDS/Base.php

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

cwolfebsco committed:
Update module/VuFindSearch/src/VuFindSearch/Backend/EDS/Base.php

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

cwolfebsco committed:
Update module/VuFindSearch/src/VuFindSearch/Backend/EDS/Base.php

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

cwolfebsco committed:
Merge branch 'EDS-API-Token-support' of https://github.com/cwolfebsco/vufind into EDS-API-Token-support

maccabeelevine commented on config/vufind/EDS.ini:
```diff
@@ -279,6 +279,12 @@ password  = "PASSWORD"
 ;password_file = /path/to/secret
 profile   = "PROFILE"
 organization_id = "VuFind from MyUniversity"
+; EBSCO might issue you with an API Key, fill in below if it was issued
+; depending on the key vended to you (legacy / new) this API key might be required
+; for successful API queries
+api_key = ""
```
I think the context of these keys is complex enough that EDS.ini needs more of this info you shared in the PR description, or else maybe a link to an 'EDS' page on the VuFind wiki where it's explained in full.

> API keys are used to monitor access to EDS API, isolating traffic by customer and use-case. This isolation allows EBSCO to manage performance for specific keys (use cases). Should a request be throttled, the gateway will respond with an HTTP 429.
Existing implementations may start using API keys at any time. EBSCO will require API keys for all customers at one point, but deadlines have not been established. Customers may have more than one API key for separate client applications or use-cases accessing EDS API.
Note: API keys do not replace EDS API’s authentication or session tokens.

maccabeelevine reviewed (changes requested):
I added some minor thoughts, but mostly I think this makes sense.

Maybe off-topic, but I was in a meeting recently with EBSCO folks who mentioned that in addition to this api key, there is the possibility of sending the user's IP address along in the EDS API request, for further specific rate limiting behavior.  I may be getting that wrong, but if so, should that be another PR or part of this one?

cwolfebsco committed:
fix: avoid disclosing full API key by mistake

cwolfebsco committed:
adding more context to EDS.ini

cwolfebsco commented:
> Maybe off-topic, but I was in a meeting recently with EBSCO folks who mentioned that in addition to this api key, there is the possibility of sending the user's IP address along in the EDS API request, for further specific rate limiting behavior. I may be getting that wrong, but if so, should that be another PR or part of this one?

You are right, this is in the plan and I was thinking of making this a second PR after seeking a discussion with you all, as the IP Address would be considered PII by some clients and thus it felt to me like it needed more thought and discussion prior to implementation.

demiankatz commented:
> You are right, this is in the plan and I was thinking of making this a second PR after seeking a discussion with you all, as the IP Address would be considered PII by some clients and thus it felt to me like it needed more thought and discussion prior to implementation.

I agree that a second PR for IP reporting would make sense, since it's a significant separate concern from the API keys and may require separate discussion. No point in holding this up for that! I'd also suggest when we get that far that it should be an "opt-in" feature due to the privacy concerns you raise. Something like `shareUserIpWithEDS = false` that can be set to true to turn on the sharing. The .ini file should have a detailed comment explaining why this might be useful, and perhaps linking to any relevant data policies.

cwolfebsco committed:
fix: moving Guest Key out of the main api key logic

While unlikely, an edge case could exist in which a customer only supports Guest Access and thus would only want to have an API Key for guests configured in EDS.ini

cwolfebsco commented:
I had a chat with Matt and we could see an "edge" case where a client would only offer Guest Access and as such I have updated Base.php accordingly. I am fairly convinced that it won't happen often, but if possible, we should support it.

I also discussed with Matt about how we plan to introduce keys. When a client will request a key, we will initially give them what we consider a "legacy" key, which is optional. once the key is sent consistently, we would make that key "new" and thus required.

demiankatz committed:
Eliminate unnecessary whitespace change.

demiankatz reviewed (approved):
Thanks, @cwolfebsco. I just pushed up a tiny change to eliminate a whitespace change in a file that was otherwise unchanged here (to simplify the PR). Other than that, it looks ready to go from my perspective, though of course I'm not currently in a position to try it hands-on.

In any case, if you've had it tested to your satisfaction, feel free to take it out of draft mode and I can merge it for you.

cwolfebsco commented:
Thanks Demian - I will take it through its paces tomorrow morning and then submit it, if it works as expected. I'll put in that nasty "debug" again so I can confirm the x-api-key values and also check with Matt to see if I can peek at the logs (or get him to peek for me). Appreciate the combine help of yourself & Maccabee

demiankatz commented:
Thanks, @cwolfebsco, sounds good! I've also requested a fresh review from @maccabeelevine since he'll have to clear his "request changes" review before we can merge.

maccabeelevine reviewed (approved):
Yup, I haven't actually tested it (don't have the api keys) but code looks good.

cwolfebsco commented:
Good morning,

I did tests this morning and keys are correctly swapped in and out - but since I did my testing between several meetings, I will do more testing either tonight or tomorrow.

demiankatz commented:
Thanks, @cwolfebsco and @maccabeelevine! I'll stay tuned for this to come out of draft; just let me know when it's time to merge!

cwolfebsco commented:
I tested this this morning with two API Profiles.

**Account 1:** 
- has both api_key and api_key_guest

**Account 2:** 
- has neither

**Account 1 - both keys configured:**
- Guest User:    api_key_guest used
- Logged in:      api_key used

**Account 1- only api_key configured:**
- Guest User:   api_key used (as expected)
- Logged in:     api_key used

**Account 2- two keys configured:**
- Receive http 154 exception: as keys do not match account (as expected)

**Account 2 - no keys configured**
- Guest & Logged in: functioned as expected

**Account 2 - empty keys configured**
- Guest & Logged in: functioned as expected

Based on these test results, I am marking this ready for review

demiankatz commented:
Thanks, it seems that the time to merge has arrived!
